src/main/java/com/webank/weid/service/impl/EvidenceServiceImpl.java

```diff
@@ -249,6 +254,62 @@ private String getSignerFromPrivKey(List<String> signers, String privateKey) {
         }
     }
 
+    /* (non-Javadoc)
+     * @see com.webank.weid.rpc.generateHash
+     * #generateHash(T object)
+     */
+    @Override
+    public <T> ResponseData<HashString> generateHash(T object) {
+        if (object == null) {
+            return new ResponseData<>(null, ErrorCode.ILLEGAL_INPUT);
+        }
+        if (object instanceof Hashable) {
+            ResponseData<String> hashResp = getHashValue((Hashable) object);
+            if (StringUtils.isEmpty(hashResp.getResult())) {
+                return new ResponseData<>(null, hashResp.getErrorCode(),
+                    hashResp.getErrorMessage());
+            }
+            return new ResponseData<>(new HashString(hashResp.getResult()), ErrorCode.SUCCESS);
+        }
+        if (object instanceof File) {
+            // This will convert all types of file into String stream
+            String rawData = convertFileToString((File) object);
+            if (StringUtils.isEmpty(rawData)) {
+                logger.error("Failed to convert file into String: {}", ((File) object).getName());
+                return new ResponseData<>(null, ErrorCode.ILLEGAL_INPUT);
+            }
+            return new ResponseData<>(new HashString(DataToolUtils.sha3(rawData)),
+                ErrorCode.SUCCESS);
+        }
+        if (object instanceof String) {
+            if (StringUtils.isEmpty((String) object)) {
+                logger.error("Input String is blank, ignored..");
+                return new ResponseData<>(null, ErrorCode.ILLEGAL_INPUT);
+            }
+            return new ResponseData<>(new HashString(DataToolUtils.sha3((String) object)),
+                ErrorCode.SUCCESS);
+        }
+        // last-resort -- use built-in method and default serializer
+        String serialized;
+        try {
+            serialized = DataToolUtils.serialize(object);
+        } catch (Exception e) {
+            logger.info("Failed to serialize object - use built-in hashCode for {}",
+                object.getClass().getCanonicalName());
+            serialized = String.valueOf(object.hashCode());
```

Comment:
We can't use hashCode here due to the following description of this function that I found in Java API document:
[Whenever it is invoked on the same object more than once during an execution of a Java application, the hashCode method must consistently return the same integer, provided no information used in equals comparisons on the object is modified. This integer need not remain consistent from one execution of an application to another execution of the same application.](https://docs.oracle.com/javase/7/docs/api/java/lang/Object.html)

Reply:
good find! Thanks @aleexzhang . We have removed this part and gone in a conservative approach.